src/token/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo

```diff
@@ -0,0 +1,199 @@
+    }
+
+    #[embeddable_as(ERC721EnumerableImpl)]
+    impl ERC721Enumerable<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl ERC721: ERC721Component::HasComponent<TContractState>,
+        +ERC721Component::ERC721HooksTrait<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of IERC721Enumerable<ComponentState<TContractState>> {
+        ///
+        fn total_supply(self: @ComponentState<TContractState>) -> u256 {
+            self.ERC721Enumerable_all_tokens_len.read()
+        }
+
+        ///
+        fn token_of_owner_by_index(
+            self: @ComponentState<TContractState>, address: ContractAddress, index: u256
+        ) -> u256 {
+            let erc721_component = get_dep_component!(self, ERC721);
+            assert(index < erc721_component.balance_of(address), Errors::OUT_OF_BOUNDS_INDEX);
+            self.ERC721Enumerable_owned_tokens.read((address, index))
+        }
+
+        ///
+        fn token_by_index(self: @ComponentState<TContractState>, index: u256) -> u256 {
+            assert(index < self.total_supply(), Errors::OUT_OF_BOUNDS_INDEX);
+            self.ERC721Enumerable_all_tokens.read(index)
+        }
+    }
+
+    #[embeddable_as(ERC721EnumerableCamelImpl)]
+    impl ERC721EnumerableCamel<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl ERC721: ERC721Component::HasComponent<TContractState>,
+        +ERC721Component::ERC721HooksTrait<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of IERC721EnumerableCamel<ComponentState<TContractState>> {
+        ///
+        fn totalSupply(self: @ComponentState<TContractState>) -> u256 {
+            self.total_supply()
+        }
+
+        ///
+        fn tokenOfOwnerByIndex(
+            self: @ComponentState<TContractState>, address: ContractAddress, index: u256
+        ) -> u256 {
+            self.token_of_owner_by_index(address, index)
+        }
+
+        ///
+        fn tokenByIndex(self: @ComponentState<TContractState>, index: u256) -> u256 {
+            self.token_by_index(index)
+        }
+    }
+
+    //
+    // Internal
+    //
+
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl ERC721: ERC721Component::HasComponent<TContractState>,
+        +ERC721Component::ERC721HooksTrait<TContractState>,
+        impl SRC5: SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        ///
+        fn initializer(ref self: ComponentState<TContractState>) {
+            let mut src5_component = get_dep_component_mut!(ref self, SRC5);
+            src5_component.register_interface(interface::IERC721ENUMERABLE_ID);
+        }
+
+        ///
+        fn _update(ref self: ComponentState<TContractState>, to: ContractAddress, token_id: u256,) {
```

Comment:
Should we provide this as a hooks implementation (after_update)? Something like a function maybe is enough, and then users can call each extension after_update in the preset Hooks implementation. We need to document it, but I think it will be easy enough to use.

Reply:
That's basically the idea behind the design. I went back and forth on `update` vs `before_update`. I do like the idea of having a pattern of extension functions (required in hooks) named `before/after_update`. Will update